lib/make-issue-token-route.ts

```diff
@@ -1,18 +1,20 @@
-import { ResponseToolkit, Request } from "@hapi/hapi";
+import { ResponseToolkit, Request, ResponseObject } from "@hapi/hapi";
+import { Boom, Payload, Output } from "@hapi/boom";
 import querystring from "querystring";
 import Wreck from "@hapi/wreck";
 import btoa from "btoa";
 import { ClientSecrets, TokenPayload } from ".";
 
-type RequestWithPossibleErrorResponse = Request & {
-  response: {
-    isBoom?: boolean;
-    data?: object;
-    output?: {
-      payload?: { [key: string]: any };
-    };
-  };
-};
+type OidcErrorData = { payload: Buffer };
+type OidcError = Required<Boom<OidcErrorData>> & { output: OidcErrorOutput };
+type OidcErrorPayload = Payload & { oidc_error: Record<string, unknown> };
+type OidcErrorOutput = Output & { payload: OidcErrorPayload };
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+type HapiResponse = ResponseObject | Required<Boom<any>>;
+function isOidcError(response: HapiResponse): response is OidcError {
+  return (response as OidcError).isBoom !== undefined;
+}
```

Comment:
One of the linter rules disallows the implicit assignment of `any` types. Hapi does nothing but declare and return `any` types in it's request and response objects.

Grant told me about [Type Guards](https://www.typescriptlang.org/docs/handbook/advanced-types.html#type-guards-and-differentiating-types) so the technique I'm using, that seems pretty good, is to declare a custom error type and use the type predicate to say "is this response object an Oidc Error?". Then typescript stops yelling at me because I told it that the things I want are there.

Reply:
Type guards are a new concept for me, that's pretty slick.